src/include/detail/graph/greedy_search.h

```diff
@@ -0,0 +1,550 @@
+    }
+
+    // q2 <- L \ V
+    // @todo Is there a better way to do this?  By somehow removing
+    // elements from q1?  We can probably make the insertion into q2
+    // more efficient -- have a batch interface to insert into q2
+    for (auto&& [s, p] : result) {
+      if (!visited(p)) {
+        q2.insert(s, p);
+      }
+    }
+
+    // L <- L \cup Nout(p*)  ; L \ V <- L \ V \cup Nout(p*)
+
+    // In looking at profiling, a majority of time is spent in this loop,
+    // in visited() and in result.insert()
+    for (auto&& [_, p] : graph.out_edges(p_star)) {
+      // assert(p != p_star);
+      if (!visited(p)) {
+        auto score = distance(db[p], query);
+
+        // unique id or not does not seem to make a difference
+        // @todo (actually it does, but shouldn't need it -- need to
+        // investigate) if (result.template insert /*<unique_id>*/ (score, p)) {
+        if (result.template insert<unique_id>(score, p)) {
+          q2.template insert<unique_id>(score, p);
+        }
+      }
+    }
+
+    if (noisy) {
+      debug_min_heap(result, "result, aka Ell: ", 1);
+      debug_min_heap(result, "result, aka Ell: ", 0);
+    }
+
+    q1.swap(q2);
+    q2.clear();
+  }
+
+  // auto top_k = Vector<id_type>(k_nn);
+  // auto top_k_scores = Vector<score_type>(k_nn);
+  auto top_k = std::vector<id_type>(k_nn);
+  auto top_k_scores = std::vector<score_type>(k_nn);
+
+  get_top_k_with_scores_from_heap(result, top_k, top_k_scores);
+  return std::make_tuple(
+      std::move(top_k_scores), std::move(top_k), std::move(visited_vertices));
+}
+
+template </* SearchPath SP, */ class Distance = sum_of_squares_distance>
+auto greedy_search(
+    auto&& graph,
+    auto&& db,
+    typename std::decay_t<decltype(graph)>::id_type source,
+    auto&& query,
+    size_t k_nn,
+    size_t L,
+    Distance&& distance = Distance{}) {
+  return greedy_search_O1(
+      std::forward<decltype(graph)>(graph),
+      std::forward<decltype(db)>(db),
+      source,
+      std::forward<decltype(query)>(query),
+      k_nn,
+      L,
+      std::forward<Distance>(distance));
+}
+
+/**
+ * @brief RobustPrune(p, vee, alpha, R)
+ * @tparam I index type
+ * @tparam Distance distance functor
+ * @param graph Graph
+ * @param p point \in P
+ * @param V candidate set
+ * @param alpha distance threshold >= 1
+ * @param R Degree bound
+ *
+ * From the DiskANN paper:
+ * V <- (V \cup Nout(p) \ p
+ * Nout(p) < 0
+ * while (!V.empty()) {
+ *  Find p* \in V with smallest distance to p
+ *  Nout(p) <- Nout(p) \cup p*
+ *  if size(Nout(p)) == R {
+ *    break
+ *  }
+ *  for pp \in V {
+ *  if alpha * distance(p*, pp) <= distance(p, pp) {
+ *    remove pp from V
+ *  }
+ * }
+ */
+template <class I = size_t, class Distance = sum_of_squares_distance>
+auto robust_prune(
+    auto&& graph,
+    auto&& db,
+    I p,
+    auto&& V_in,
+    float alpha,
+    size_t R,
+    Distance&& distance = Distance{}) {
+  constexpr bool noisy = false;
```

Comment:
(Minor) Should this use the top-level `noisy`?

Reply:
Created a separate `noisy_robust_prune` to separate noisy `greedy_search` from `robust_prune`